### PR TITLE
Development dashboard

### PR DESCRIPTION
**Pull request by ensist**
...

### CODE REVIEW

File: model/pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
     <parent>
         <groupId>com.oasis</groupId>
         <artifactId>oasis</artifactId>
-        <version>1.0.0-SNAPSHOT</version>
+        <version>1.1.0-SNAPSHOT</version>
```
Comment (ensist) [inline model/pom.xml:8]: no need to change the version, you can still use the old version

File: model/src/main/java/com/oasis/model/FieldName.java [hunk 2]
```diff
@@ -11,4 +11,24 @@
     public static final String EMPLOYEE_DIVISION = "division";
     public static final String EMPLOYEE_SUPERVISING_COUNT = "supervisingCount";
     public static final String EMPLOYEE_SUPERVISION_ID = "supervisionId";
+
+    //Field names for AssetModel
+    public static final String ASSET_NAME = "name";
```
Comment (ensist) [inline model/src/main/java/com/oasis/model/FieldName.java]: separate it to another classes
AssetFieldName, RequestFieldName, etc

File: model/src/main/java/com/oasis/model/entity/AssetModel.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@Document(collection = CollectionName.ASSET_COLLECTION_NAME)
+public class AssetModel extends BaseEntity {
+    @Field(FieldName.ASSET_NAME)
+    private String name;
+
+    @Field(FieldName.ASSET_LOCATION)
+    private String location;
+
+    @Field(FieldName.ASSET_PRICE)
+    private Double price;
```
Comment (ensist) [inline model/src/main/java/com/oasis/model/entity/AssetModel.java]: price should not be null, use primitive data type double

File: model/src/main/java/com/oasis/model/entity/AssetModel.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Field(FieldName.ASSET_STOCK)
+    private Integer stock;
```
Comment (ensist) [inline model/src/main/java/com/oasis/model/entity/AssetModel.java]: use primitive int

File: pom.xml [hunk 5]
```diff
@@ -23,13 +23,18 @@
...
 
-        <oasis.version>1.0.0-SNAPSHOT</oasis.version>
         <springwebmvc.version>5.1.0.RELEASE</springwebmvc.version>
         <springcontext.version>5.0.8.RELEASE</springcontext.version>
         <springsecurity.version>3.1.0.RELEASE</springsecurity.version>
         <mongodb.version>2.1.0.RELEASE</mongodb.version>
         <jackson.version>2.9.4</jackson.version>
         <lombok.version>1.18.2</lombok.version>
+        <orika.version>1.5.2</orika.version>
+        <model.version>${project.parent.version}</model.version>
```
Comment (ensist) [inline pom.xml]: no need to separate model.version, repository.version, etc.
all of those will always have same version

File: repository/src/main/java/com/oasis/repository/AssetRepository.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.oasis.model.entity.AssetModel;
+import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface AssetRepository extends MongoRepository<AssetModel, String> {
+    List<AssetModel> findAllByStockGreaterThan(Integer stockLimit);
```
Comment (ensist) [inline repository/src/main/java/com/oasis/repository/AssetRepository.java]: add a new line here

Comment (ensist) [reply repository/src/main/java/com/oasis/repository/AssetRepository.java]: use primitive int

File: repository/src/main/java/com/oasis/repository/AssetRepository.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    AssetModel findBy_id(String assetId);
```
Comment (ensist) [inline repository/src/main/java/com/oasis/repository/AssetRepository.java:12]: I think it should be findById

File: repository/src/main/java/com/oasis/repository/RequestRepository.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+import com.oasis.model.entity.RequestModel;
+import org.springframework.data.mongodb.repository.MongoRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface RequestRepository extends MongoRepository<RequestModel, String> {
+    List<RequestModel> findAllByStatusAndEmployeeId(String status, String employeeId);
```
Comment (ensist) [inline repository/src/main/java/com/oasis/repository/RequestRepository.java]: switch the parameter, employeeId first then status.
prefer to read it from the most unique one
findAllByEmployeeIdAndStatus

File: service/src/main/java/com/oasis/service/RoleDeterminer.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        EmployeeModel employee = employeeRepository.findBy_id(employeeId);
+
+        if (employee.getSupervisingCount() == 0) {
+            return ROLE_EMPLOYEE;
+        } else {
+            List<SupervisionModel> supervisions = supervisionRepository.findAllBySupervisorId(employeeId);
+            for (SupervisionModel supervision : supervisions) {
+                EmployeeModel supervisedEmployee = employeeRepository.findBy_id(supervision.getEmployeeId());
+                if (supervisedEmployee.getSupervisingCount() > 0) {
+                    return ROLE_ADMINISTRATOR;
```
Comment (ensist) [inline service/src/main/java/com/oasis/service/RoleDeterminer.java:25]: wrong logic for administrator, just add a new field isAdmin in employee collection or create a new table to store admin username and password

File: service/src/main/java/com/oasis/service/ServiceConstant.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.oasis.service;
+
+public class ServiceConstant {
+    public static final Integer STOCK_LIMIT = 0;
```
Comment (ensist) [inline service/src/main/java/com/oasis/service/ServiceConstant.java]: Rename to `ZERO`

File: service/src/main/java/com/oasis/service/api/DashboardServiceApi.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.oasis.model.entity.AssetModel;
+import com.oasis.model.entity.EmployeeModel;
+import com.oasis.model.entity.RequestModel;
+import com.oasis.model.entity.SupervisionModel;
+
+import java.util.List;
+
+public interface DashboardServiceApi {
+    Integer getAvailableAssetsCount();
```
Comment (ensist) [inline service/src/main/java/com/oasis/service/api/DashboardServiceApi.java]: add a new line below method declaration

File: web-controller/src/main/java/com/oasis/constant/ControllerConstant.java [hunk 12]
```diff
@@ -0,0 +1,5 @@
+package com.oasis.constant;
+
+public class ControllerConstant {
+    public static final Integer DASHBOARD_REQUEST_UPDATE_PAGE_SIZE = 5;
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/constant/ControllerConstant.java]: change to primitive int

File: web-controller/src/main/java/com/oasis/constant/ErrorCodeAndMessage.java [hunk 13]
```diff
@@ -7,4 +7,10 @@
     public static final String[] PASSWORD_DOES_NOT_MATCH = {
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/constant/ErrorCodeAndMessage.java]: change from array of string to object
ErrorCode {
String code;
String message;
}

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;
+
+@CrossOrigin(origins = "http://localhost")
+@RestController
+public class DashboardController {
+    @Autowired
+    private DashboardServiceImpl dashboardServiceImpl;
+
+    @GetMapping(value = APIMappingValue.API_DASHBOARD_STATUS,
+            produces = APPLICATION_JSON_VALUE)
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: add consumes

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+        sortData(requestUpdates, sortInfo);
+        List<RequestUpdateModel> mappedRequests = mapRequests(requestUpdates);
+
+        return produceDashboardRequestUpdateSuccessResult(
+                mappedRequests,
+                pageNumber
+        );
+    }
+
+    private NoPagingResponse<DashboardStatusResponse>
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: separate it to different class

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+                        requestedRequestsCount,
+                        pendingHandoverRequestsCount,
+                        availableAssetCount
+                )
+        );
+
+        return successResponse;
+    }
+
+    private PagingResponse<DashboardRequestUpdateResponse>
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: separate it to different class

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+                        pageNumber,
+                        ControllerConstant.DASHBOARD_REQUEST_UPDATE_PAGE_SIZE,
+                        requests.size()
+                )
+        );
+
+        return successResponse;
+    }
+
+    private PagingResponse<FailedResponse>
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: separate it to different class

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+                        errorCode,
+                        errorMessage
+                )
+        );
+        failedResponse.setPaging(null);
+
+        return failedResponse;
+    }
+
+    private void fillData(List<RequestModel> requestUpdates, String employeeId, String currentTab, String role) {
```
Comment (ensist) [inline web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: separate it to different class

File: model/src/main/java/com/oasis/model/FieldName.java (hunk 2 above)
Comment (jonathan016) [reply model/src/main/java/com/oasis/model/FieldName.java]: Commit ID: 57faedf

File: service/src/main/java/com/oasis/service/ServiceConstant.java (hunk 10 above)
Comment (jonathan016) [reply service/src/main/java/com/oasis/service/ServiceConstant.java]: Commit ID: 5a96711

File: repository/src/main/java/com/oasis/repository/AssetRepository.java (hunk 6 above)
Comment (jonathan016) [reply repository/src/main/java/com/oasis/repository/AssetRepository.java]: Commit ID: 5a96711

File: model/src/main/java/com/oasis/model/entity/AssetModel.java (hunk 4 above)
Comment (jonathan016) [reply model/src/main/java/com/oasis/model/entity/AssetModel.java]: Commit ID: 5a96711

File: model/src/main/java/com/oasis/model/entity/AssetModel.java (hunk 3 above)
Comment (jonathan016) [reply model/src/main/java/com/oasis/model/entity/AssetModel.java]: Commit ID: 5a96711

File: web-controller/src/main/java/com/oasis/constant/ControllerConstant.java (hunk 12 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/constant/ControllerConstant.java]: Commit ID: 5a96711

File: repository/src/main/java/com/oasis/repository/AssetRepository.java (hunk 6 above)
Comment (jonathan016) [reply repository/src/main/java/com/oasis/repository/AssetRepository.java]: Where should the new line be Kak? After line 10 or 11?

File: service/src/main/java/com/oasis/service/api/DashboardServiceApi.java (hunk 11 above)
Comment (jonathan016) [reply service/src/main/java/com/oasis/service/api/DashboardServiceApi.java]: Commit ID: a81d3e2

File: repository/src/main/java/com/oasis/repository/RequestRepository.java (hunk 8 above)
Comment (jonathan016) [reply repository/src/main/java/com/oasis/repository/RequestRepository.java]: Commit ID: a81d3e2

File: pom.xml (hunk 5 above)
Comment (jonathan016) [reply pom.xml]: Commit ID: a81d3e2

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java (hunk 14 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: Commit ID: 4a81232

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java (hunk 15 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: Commit ID: 4a81232

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java (hunk 16 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: Commit ID: 4a81232

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java (hunk 17 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: Commit ID: 4a81232

File: web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java (hunk 18 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/webcontroller/DashboardController.java]: Commit ID: 4a81232

File: web-controller/src/main/java/com/oasis/constant/ErrorCodeAndMessage.java (hunk 13 above)
Comment (jonathan016) [reply web-controller/src/main/java/com/oasis/constant/ErrorCodeAndMessage.java]: Commit ID: 4a81232 (if not mistaken)

File: repository/src/main/java/com/oasis/repository/AssetRepository.java (hunk 7 above)
Comment (jonathan016) [reply repository/src/main/java/com/oasis/repository/AssetRepository.java:12]: This is fixed in assets feature

**Pull request merged**